### DISCUSSION TITLE
Does buzzer support work?

### DISCUSSION THREAD

**Discussion by JJPro** [emporia-vue-local/esphome#150] [Q&A]
Buzzer support has been added since Feb update. It never worked for me. 
I'm using the default buzzer song shown in README file. 

Does buzzer work for you guys, or does my unit have a bad buzzer?

**Comment by rohankapoorcom** [comment]
Does not work on any of my three units at all. Hoping there's a fix!

**Comment by bry25uk** [comment]
DOESNT WORK ON MINE v2

**Comment by jklap** [comment]
Anyone have it open and could run a continuity test to determine which pin the buzzer is connected to?

**Comment by FirbyKirby** [comment]
I can also confirm no buzzer is working after adding the buzzer support from the latest configuration example.

@pgwipeout , do you have any thoughts or suggestions since, if I'm not mistaken, you authored the pull request that added buzzer support? And I'll take a second to thank you for your contribution at the same time. The updates to the style and great, and the LED support works great.

**Comment by pgwipeout** [comment]
The RTTL based buzzer is extremely quiet. I know the buzzer is capable of screaming as I've made the original firmware do it during testing, however I haven't found an esphome configuration that could replicate that. It's loud enough to make out when outside the panel, but inside I can only make it out when playing long strings and the house is quiet. I added the buzzer mostly to document the pin the buzzer lives on. It needs a PWM signal to generate any sound, as it's a piezo buzzer directly connected to a gpio.

**Reply by pgwipeout** [reply]
I've checked into this, and it seems that yes at some point it was broken in newer versions of esphome.

**Reply by g3gg0** [reply]
It's because of an incorrect GPIO setup.
you have to set IO27 to output and set it to low - IO12 is for the audio signal

**Reply by pgwipeout** [reply]
Confirmed, IO27 must be set as output. Funnily enough, setting it to low or high doesn't seem to affect it much (though I suspect it will invert the PWM strength).

**Reply by pgwipeout** [reply]
Swapping 12 and 27 also works. I wonder if the stock firmware drove each pin with an opposite waveform to generate the extremely loud output.

**Reply by g3gg0** [reply]
yep, the buzzer has a polarity IO27 GND and IO12 PWM.
however, driving them inverted would probably increase loudness.
i am using the LEDC and did not see anything like that in the espressif API.
(did not check the datasheet yet, if the peripheral supports inverted outputs)

**Comment by agoode** [comment]
This made my buzzer work properly:

```
output:
  - platform: ledc
    pin: GPIO12
    id: buzzer
  - platform: gpio
    pin: GPIO27
    id: buzzer_gnd
```

**Reply by FirbyKirby** [reply]
Did you add anyting else when you added the additional platform declaration to the output stanza? I replicated your setup:
```
output:
  # set up the rtttl buzzer on GPIO12 using a PWM output (LEDC.) Set GPIO27 to GND as it's attached to the buzzer and provides a GND path.
  - platform: ledc
    pin: GPIO12
    id: buzzer
  - platform: gpio
    pin: GPIO27
    id: buzzer_gnd
```

And am getting a compile error upon deployment.

```
INFO ESPHome 2023.5.5
INFO Reading configuration /config/esphome/emporiavue2-main.yaml...
WARNING GPIO12 is a Strapping PIN and should be avoided.
Attaching external pullup/down resistors to strapping pins can cause unexpected failures.
See https://esphome.io/guides/faq.html#why-am-i-getting-a-warning-about-strapping-pins
INFO Detected timezone 'America/Chicago'
INFO Generating C++ source...
INFO Compiling app...
Processing emporiavue2-main (board: esp32dev; framework: espidf; platform: platformio/espressif32@5.3.0)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.40404.0 (4.4.4) 
 - tool-cmake @ 3.16.4 
 - tool-ninja @ 1.7.1 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-xtensa-esp32 @ 8.4.0+2021r2-patch5
Reading CMake configuration...
Dependency Graph
|-- Improv @ 1.2.3
|-- noise-c @ 0.1.4
|   |-- libsodium @ 1.10018.1
Compiling /data/emporiavue2-main/.pioenvs/emporiavue2-main/src/main.o
Linking /data/emporiavue2-main/.pioenvs/emporiavue2-main/firmware.elf
/data/cache/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /data/emporiavue2-main/.pioenvs/emporiavue2-main/src/main.o:(.literal._Z5setupv+0xc4): undefined reference to `vtable for esphome::gpio::GPIOBinaryOutput'
/data/cache/platformio/packages/toolchain-xtensa-esp32/bin/../lib/gcc/xtensa-esp32-elf/8.4.0/../../../../xtensa-esp32-elf/bin/ld: /data/emporiavue2-main/.pioenvs/emporiavue2-main/src/main.o:(.literal._Z5setupv+0xc8): undefined reference to `vtable for esphome::gpio::GPIOBinaryOutput'
collect2: error: ld returned 1 exit status
*** [/data/emporiavue2-main/.pioenvs/emporiavue2-main/firmware.elf] Error 1
========================= [FAILED] Took 49.10 seconds =========================
```

I'm not familiar with the root error `undefined reference to `vtable for esphome::gpio::GPIOBinaryOutput'` but if I had to speculate, it seems the GPIO platform isn't linked in properly, or perhaps the pin assignment is incorrect. Not much else is getting configured, honestly.

I believe I've got an updated ESPHome version (2023.5.5).

**Reply by rohankapoorcom** [reply]
I've noticed with this project, I need to use the "Clean Build Files" every time I need to build. Running that and then deploying again fixed it for me.

**Reply by FirbyKirby** [reply]
> I've noticed with this project, I need to use the "Clean Build Files" every time I need to build. Running that and then deploying again fixed it for me.

That did the trick! I didn't even know the "Clean Build Files" option existed. Thanks!
